### ISSUE TITLE
salt-cloud CLI 2015.8 borks out with SaltClientError: 'timeout'

### ISSUE THREAD

**Issue by The-Loeki** [saltstack/salt#27960]
During deploying a single VM with the `cloudstack` driver with `salt-cloud` CLI, it borks out:

```
[ERROR   ] There was a profile error: 'timeout'
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/cloud/cli.py", line 276, in run
    self.config.get('names')
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1405, in run_profile
    ret[name] = self.create(vm_)
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1273, in create
    self.opts['sync_after_install']
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 542, in cmd
    **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 297, in run_job
    raise SaltClientError(general_exception)
SaltClientError: 'timeout'
```

(and yes, considering #27953 and #27956, this is turning out to be a quite bumpy ride :P)

**Comment by rallytime** [comment]
I just asked @techhat about this. His on the case!

**Comment by techhat** [comment]
@basepi, I'm not sure if this is a cloud or a core issue. Would you please take a look at it?

**Comment by basepi** [comment]
Looks like it's probably core, technically. May be fixable just by raising the main timeout, or the secondary timeout (`gather_job_timeout`).

**Comment by The-Loeki** [comment]
@basepi FYI & roughly from the top of my head (I can provide more precise timings if you want)

* from start (`salt-cloud -p profile oneminion`) to trace is ~1min
* from sending the `created instance` event to trace is <10sec

```
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/minion/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/minion/master_event_pull.ipc
[ERROR   ] There was a profile error: 'timeout'
(stacktrace)
```
 between these three is <1sec

**Comment by cachedout** [comment]
Can you recreate this problem with master debugging turned on and then paste those logs? `run_job` should be non-blocking and this stacktrace is suggesting that the master just completely went away.

**Comment by The-Loeki** [comment]
@cachedout of course, but I'm not sure what you're hoping to gain, as the messages are the same:

```
2015-10-20 14:16:59,136 [salt.utils.event ][DEBUG   ][4687] MasterEvent PUB socket URI: ipc:///var/run/salt/master/master_event_pub.ipc
2015-10-20 14:16:59,137 [salt.utils.event ][DEBUG   ][4687] MasterEvent PULL socket URI: ipc:///var/run/salt/master/master_event_pull.ipc
2015-10-20 14:16:59,137 [salt.utils.event ][DEBUG   ][4687] Sending event - data = {'profile': 'aurora_agile_centos_1g', 'event': 'created instance', '_stamp': '2015-10-20T12:16:59.137158', 'name': 'dodacloud02', 'provider': 'aurora-ams:cloudstack'}
2015-10-20 14:17:02,194 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion
2015-10-20 14:17:02,293 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/_schedule.conf'
2015-10-20 14:17:02,294 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/_schedule.conf
2015-10-20 14:17:02,298 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/cloudmonkey.conf'
2015-10-20 14:17:02,299 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/cloudmonkey.conf
2015-10-20 14:17:02,301 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/ipv6.conf'
2015-10-20 14:17:02,302 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/ipv6.conf
2015-10-20 14:17:02,303 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/reconnect-timings.conf'
2015-10-20 14:17:02,304 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/reconnect-timings.conf
2015-10-20 14:17:02,329 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/restart-on-error.conf'
2015-10-20 14:17:02,330 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/restart-on-error.conf
2015-10-20 14:17:02,332 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/start-with-states.conf'
2015-10-20 14:17:02,332 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/start-with-states.conf
2015-10-20 14:17:02,335 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/state-aggregrate.conf'
2015-10-20 14:17:02,335 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/state-aggregrate.conf
2015-10-20 14:17:02,337 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/use-ipv6.conf'
2015-10-20 14:17:02,338 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/use-ipv6.conf
2015-10-20 14:17:02,341 [salt.config      ][DEBUG   ][4687] Including configuration from '/etc/salt/minion.d/use-requests.conf'
2015-10-20 14:17:02,341 [salt.config      ][DEBUG   ][4687] Reading configuration from /etc/salt/minion.d/use-requests.conf
2015-10-20 14:17:02,347 [salt.utils.event ][DEBUG   ][4687] MasterEvent PUB socket URI: ipc:///var/run/salt/minion/master_event_pub.ipc
2015-10-20 14:17:02,348 [salt.utils.event ][DEBUG   ][4687] MasterEvent PULL socket URI: ipc:///var/run/salt/minion/master_event_pull.ipc
2015-10-20 14:17:02,381 [salt.cloud.cli   ][ERROR   ][4687] There was a profile error: 'timeout'
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/cloud/cli.py", line 276, in run
    self.config.get('names')
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1403, in run_profile
    ret[name] = self.create(vm_)
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1271, in create
    self.opts['sync_after_install']
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 542, in cmd
    **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 297, in run_job
    raise SaltClientError(general_exception)
SaltClientError: 'timeout'
```

**Comment by cachedout** [comment]
Wow, interesting. OK. I'll have to dig deeper into the code to figure out what's going on here. @rallytime should probably get involved here too.

**Comment by The-Loeki** [comment]
OK, wait, I'll see if I can get SS to debug you more

**Comment by The-Loeki** [comment]
As a side note, this is run on the salt master

```
[DEBUG   ] Reading configuration from /etc/salt/minion.d/use-requests.conf
[DEBUG   ] MasterEvent PUB socket URI: ipc:///var/run/salt/minion/master_event_pub.ipc
[DEBUG   ] MasterEvent PULL socket URI: ipc:///var/run/salt/minion/master_event_pull.ipc
[DEBUG   ] He @cachedout! Look at my debug before running run_job: tgt: dodacloud04, fun: saltutil.sync_all, arg: [], expr_form: glob, ret: , timeout: None, jid: , kwargs: {}
[ERROR   ] Failed to create VM dodacloud04. Configuration value 'timeout' needs to be set
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1271, in create
    self.opts['sync_after_install']
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 545, in cmd
    **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 280, in run_job
    log.debug('He @cachedout! Look at my debug before running self.pub: tgt: {}, fun: {}, arg: {}, expr_form: {}, ret: {}, timeout: {}, jid: {}, kwargs: {}'.format(tgt, fun, arg, expr_form, ret, self._get_timeout(time
out), jid, kwargs))
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 187, in _get_timeout
    return self.opts['timeout']
KeyError: 'timeout'
```

**Comment by The-Loeki** [comment]
Am I right in assuming somehow there's no decent default for `timeout`?

**Comment by bl4ckcontact** [comment]
This also occurs with the `openstack` driver, if it helps any.

**Comment by cmclaughlin** [comment]
I was also seeing this booting EC2 instances with salt-cloud.  I hacked the source and bumped up the timeout really high (gather_job_timeout and the main timeout didn't seem to help).  Then I got another stacktrace and this error:

    SaltClientError: The salt master could not be contacted. Is master running?

After more trial and error, I narrowed down my problem to this salt-cloud setting:

    sync_after_install: all

I'm new to Salt and don't even have any custom modules or grains yet, but I thought that feature sounded good so I had it on.  I disabled that and now salt-cloud runs great.  Not sure if this will help anyone... but perhaps you might want to try the same thing.

Even though it was pebkac in my case, it would be nice to have better output.

**Comment by gladiatr72** [comment]
any headway on this?

**Comment by mlalpho** [comment]
I looked in to this a bit further as it still shows up for me.

Yes, @cmclaughlin is right about how to avoid the Exception. The code is called from `salt/cloud/__init__.py` only if sync_after_install is set.

At `salt/cloud/__init__.py", line 1272` the call to `client.cmd` requires timeout kwarg set
```
            if output is not False and 'sync_after_install' in self.opts:
                if self.opts['sync_after_install'] not in (
                        'all', 'modules', 'states', 'grains'):
                    log.error('Bad option for sync_after_install')
                    return output

                # a small pause makes the sync work reliably
                time.sleep(3)

                mopts_ = salt.config.DEFAULT_MINION_OPTS
                conf_path = '/'.join(self.opts['conf_file'].split('/')[:-1])
                mopts_.update(
                    salt.config.minion_config(
                        os.path.join(conf_path,
                                     'minion')
                    )
                )

                client = salt.client.get_local_client(mopts=mopts_)
                **log.debug( '{0}'.format(mopts_) )**
                ret = client.cmd(vm_['name'], 'saltutil.sync_{0}'.format(
                    self.opts['sync_after_install']
                ), **timeout=10**)
```

After we get past the 'timeout' error ..

When the process gets to `salt/client/__init__.py` `def pub`, this is where we see the connection error:
```
        # Make sure the publisher is running by checking the unix socket
        if (self.opts.get('ipc_mode', '') != 'tcp' and
                not os.path.exists(os.path.join(self.opts['sock_dir'],
                'publish_pull.ipc'))):
            log.error(
                'Unable to connect to the salt master publisher at '
                '{0}'.format(self.opts['sock_dir'])
            )
            raise SaltClientError
```
Adding some debug statements before the block where the error is thrown:
```
log.debug( '{0}'.format(self.opts) )
log.debug( 'Looking at {1}: {0}'.format(os.path.exists(os.path.join(self.opts['sock_dir'], 'publish_pull.ipc')),self.opts['sock_dir']))
log.debug( 'Looking at /var/run/salt/master: {0}'.format(os.path.exists(os.path.join('/var/run/salt/master', 'publish_pull.ipc'))))
```
Produces:
```
{'zmq_filtering': False, 'recon_default': 1000, 'state_output_diff': False, 'fileserver_backend': ['roots'], 'environment': None, 'acceptance_wait_time': 10, 'disable_modules': [], 'backup_mode': '', 'file_ignore_glob': None, 'winrepo_ssl_verify': False, 'cluster_mode': False, 'gitfs_pubkey': '', 'fileserver_ignoresymlinks': False, 'file_ignore_regex': None, 'multiprocessing': True, 'git_pillar_insecure_auth': False, 'returner_dirs': [], 'gitfs_privkey': '', 'grains_dirs': [], 'render_dirs': [], 'gitfs_user': '', 'fileserver_limit_traversal': False, 'tcp_keepalive_intvl': -1, 'conf_file': '/etc/salt/minion', 'zmq_monitor': False, 'tcp_pull_port': 4511, 'pidfile': '/var/run/salt-minion.pid', 'log_datefmt_logfile': '%Y-%m-%d %H:%M:%S', 'winrepo_cachefile': 'winrepo.p', 'raet_main': False, 'cachedir': '/var/cache/salt/minion', 'interface': '0.0.0.0', 'recon_max': 10000, 'default_include': 'minion.d/*.conf', 'http_request_timeout': 3600.0, 'state_events': False, 'restart_on_error': False, 'winrepo_pubkey': '', 'ipc_mode': 'ipc', 'keysize': 2048, 'master_sign_key_name': 'master_sign', 'raet_port': 4510, 'git_pillar_env': '', 'hash_type': 'md5', 'state_output': 'full', 'modules_max_memory': -1, 'master_shuffle': False, 'gitfs_env_whitelist': [], 'extension_modules': '/var/cache/salt/minion/extmods', '__cli': 'salt-cloud', 'grains_cache': False, 'sudo_user': '', 'username': None, 'cmd_safe': True, 'log_fmt_logfile': '%(asctime)s,%(msecs)03.0f [%(name)-17s][%(levelname)-8s][%(process)d] %(message)s', 'enable_zip_modules': False, 'grains': {}, 'top_file_merging_strategy': 'merge', 'raet_lane_bufcnt': 100, 'git_pillar_root': '', 'winrepo_remotes_ng': ['https://github.com/saltstack/salt-winrepo-ng.git'], 'raet_alt_port': 4511, 'state_aggregate': False, 'winrepo_branch': 'master', 'providers': {}, 'winrepo_remotes': ['https://github.com/saltstack/salt-winrepo.git'], 'grains_cache_expiration': 300, 'id': 'docker-test-internal-salt-master01', 'caller_floscript': '/usr/lib/python2.7/site-packages/daemons/flo/caller.flo', 'log_level': None, 'git_pillar_password': '', 'auth_safemode': False, 'clean_dynamic_modules': True, 'disable_returners': [], 'cache_sreqs': True, 'git_pillar_pubkey': '', 'tcp_keepalive': True, 'file_client': 'remote', 'user': 'root', 'use_master_when_local': False, 'git_pillar_base': 'master', 'open_mode': False, 'gitfs_env_blacklist': [], 'winrepo_dir_ng': '/srv/salt/win/repo-ng', 'winrepo_password': '', 'pki_dir': '/etc/salt/pki/minion', 'minion_floscript': '/usr/lib/python2.7/site-packages/daemons/flo/minion.flo', 'ioflo_verbose': 0, 'sock_dir': '/var/run/salt/minion', 'tcp_pub_port': 4510, 'gitfs_remotes': [], 'git_pillar_passphrase': '', 'ioflo_realtime': True, 'master_alive_interval': 0, 'whitelist_modules': [], 'utils_dirs': ['/var/cache/salt/minion/extmods/utils'], 'state_top_saltenv': None, 'states_dirs': [], 'outputter_dirs': [], 'module_dirs': [], 'git_pillar_ssl_verify': False, 'gitfs_insecure_auth': False, 'state_auto_order': True, '__role': 'minion', 'minion_id_caching': True, 'cluster_masters': [], 'recon_randomize': True, 'auto_accept': True, 'cache_jobs': False, 'file_recv': False, 'state_verbose': True, 'verify_master_pubkey_sign': False, 'password': None, 'startup_states': '', 'retry_dns': 30, 'return_retry_timer': 5, 'pillar_roots': {'base': ['/srv/pillar', '/srv/spm/pillar']}, 'file_roots': {'base': ['/srv/salt', '/srv/spm/salt']}, 'root_dir': '/', 'log_granular_levels': {}, 'state_top': 'top.sls', 'transport': 'zeromq', 'gitfs_base': 'master', 'ioflo_console_logdir': '', 'top_file': '', 'raet_road_bufcnt': 2, 'file_recv_max_size': 100, 'range_server': 'range:80', 'raet_mutable': False, 'permissive_pki_access': False, 'schedule': {'__mine_interval': {'function': 'mine.update', 'jid_include': True, 'minutes': 60, 'maxrunning': 2}}, 'winrepo_passphrase': '', 'update_restart_services': [], 'master_uri_format': 'default', 'rejected_retry': False, 'gitfs_ssl_verify': False, 'http_max_body': 107374182400, 'winrepo_dir': '/srv/salt/win/repo', 'cython_enable': False, 'sls_list': [], 'beacons': {}, 'ext_job_cache': '', 'winrepo_source_dir': 'salt://win/repo-ng/', 'always_verify_signature': False, 'renderer': 'yaml_jinja', 'random_reauth_delay': 10, 'auth_tries': 7, 'winrepo_insecure_auth': False, 'mine_interval': 60, 'git_pillar_privkey': '', 'winrepo_user': '', 'master': 'localhost', 'log_level_logfile': None, 'ipv6': False, 'verify_env': True, 'master_finger': '', 'syndic_finger': '', 'tcp_keepalive_idle': 300, 'gitfs_passphrase': '', 'fileserver_followsymlinks': True, 'master_port': '4506', 'ioflo_period': 0.1, 'ping_interval': 0, 'auth_timeout': 60, 'tcp_keepalive_cnt': -1, 'env_order': [], 'syndic_log_file': '/var/log/salt/syndic', 'update_url': False, 'grains_refresh_every': 0, 'master_type': 'str', 'git_pillar_branch': 'master', 'default_top': 'base', 'autoload_dynamic_modules': True, 'file_buffer_size': 262144, 'log_fmt_console': '[%(levelname)-8s] %(message)s', 'random_master': False, 'log_datefmt': '%H:%M:%S', 'max_event_size': 1048576, 'syndic_pidfile': '/var/run/salt-syndic.pid', 'loop_interval': 1, 'return_retry_timer_max': 10, 'git_pillar_user': '', 'autosign_timeout': 120, 'gitfs_root': '', 'test': False, 'gitfs_password': '', 'failhard': False, 'raet_clear_remotes': True, 'raet_clear_remote_masters': True, 'pillarenv': None, 'acceptance_wait_time_max': 0, 'winrepo_privkey': '', 'log_file': '/var/log/salt/minion', 'gitfs_mountpoint': ''}
[DEBUG   ] Looking at /var/run/salt/minion: False
[DEBUG   ] Looking at /var/run/salt/master: True

```
Which looks to me like minion config options.

Changing the 'sock_dir' to `/var/run/salt/master` gets us one more step further:
```
[ERROR   ] There was a profile error: 'syndic_master'
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/salt/cloud/cli.py", line 278, in run
    self.config.get('names')
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1403, in run_profile
    ret[name] = self.create(vm_)
  File "/usr/lib/python2.7/site-packages/salt/cloud/__init__.py", line 1272, in create
    ), timeout=10)
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 543, in cmd
    **kwargs)
  File "/usr/lib/python2.7/site-packages/salt/client/__init__.py", line 298, in run_job
    raise SaltClientError(general_exception)
SaltClientError: 'syndic_master'
```
`syndic_master` is a key for master configuration ..

It seems like it should be a master configuration dictionary in `self.opts` instead of the minion config that is present, currently. It would appear that the salt-cloud client is attempting to publish an event to the new minion from the master .. but someone `self` contains the minion details instead of the master's ..

I don't have enough knowledge of the core python as someone else to say if this is the case, how or where the swap occurred, what the expected result should be or where it could be fixed.. Maybe someone could confirm what I'm seeing?

Let me know if I can be of any help with more testing or help diagnose further :)

Cheers! :+1:

**Comment by techhat** [comment]
Fixed with #31747.
